### ISSUE TITLE
Confusion around the sslCertificateArn in chapter-6 network stack

### ISSUE THREAD

**Issue by alrawas** [stratospheric-dev/stratospheric#52]
I am currently playing around with deploying chapter-6 network stack.

I see in package.json an acm certificate. It's not yet mentioned in the book where and how we can generate it.

When I tried to do so, I went to https://docs.aws.amazon.com/acm/latest/userguide/gs-acm-request-public.html#request-public-cli

Requesting a certificate requires a parameter --domain-name http://example.com 
But in the mean time I don't want to give a domain name.

I'm a bit confused about this one. Can you please give me a hint or link me somewhere useful?

can I remove the sslCertificateArn=arn:aws:acm:... altogether?

**Comment by rieckpil** [comment]
Thanks for reporting this @alrawas.

We'll work on a fix and try to include for the release this Sunday 🚀

**Comment by thombergs** [comment]
Hey @alrawas,

you should be able just to leave the `sslCertificateArn` parameter away. It's optional in the `Network` construct, so you don't have to provide it. If you don't provide it, it just means that the loadbalancer will only be available via HTTP and not via HTTPS.

Let us know if that helps!

**Comment by alrawas** [comment]
Hello @thombergs 

I did leave the `sslCertificateArn` away, the repository app and the network, both deployed successfully with no errors.

Continuing on, when deploying the service stack, I got an error about some token.

```
cdk deploy --app "./mvnw -e -q compile exec:java -Dexec.mainClass=dev.stratospheric.todoapp.cdk.ServiceApp" -c springProfile=aws -c dockerImageUrl=docker.io/stratospheric/todo-app-v1:latest --require-approval never "-c" "accountId=------------" "-c" "region=us-east-1" "-c" "environmentName=STAGING" "-c" "applicationName=hello" "-c" "springProgile=aws" "-c" "dockerImageUrl=docker.io/stratospheric/todo-app-v1:latest"

ServiceStack (STAGING-hello-Service): deploying...
STAGING-hello-Service: creating CloudFormation changeset...
[██████████████████████▎···································] (5/13)

12:15:55 PM | CREATE_FAILED        | AWS::ElasticLoadBalancingV2::ListenerRule | ServiceStack/Service/httpsListenerRule
Resource handler returned message: "Invalid request provided: AWS::ElasticLoadBalancingV2::ListenerRule Validation exception" (RequestToken: d733dc0b-c79b-da8e-5770-ab4607fb1a
2b, HandlerErrorCode: InvalidRequest)
12:15:56 PM | ROLLBACK_IN_PROGRESS | AWS::CloudFormation::Stack                | STAGING-hello-Service
The following resource(s) failed to create: [ServicehttpsListenerRule888D8DF6, ServiceecsSecurityGroup6D8FF04D, ServicehttpListenerRule46A86AF4, ServiceecsTaskExecutionRole0EB
E7C5D, ServiceecsTaskRole4E8999D2]. Rollback requested by user.
12:15:56 PM | ROLLBACK_IN_PROGRESS | AWS::CloudFormation::Stack                | STAGING-hello-Service
The following resource(s) failed to create: [ServicehttpsListenerRule888D8DF6, ServiceecsSecurityGroup6D8FF04D, ServicehttpListenerRule46A86AF4, ServiceecsTaskExecutionRole0EB
E7C5D, ServiceecsTaskRole4E8999D2]. Rollback requested by user.
```

**Comment by thombergs** [comment]
Hey @alrawas 

the "RequestToken" is just the ID of the request that the CDK sends to CloudFormation. The actual error message is this:

```
ServiceStack/Service/httpsListenerRule
Resource handler returned message: "Invalid request provided: AWS::ElasticLoadBalancingV2::ListenerRule Validation exception"
```

The issue still has to do with the SSL configuration, I believe. CloudFormation is complaining that the `httpsListener` is somehow invalid (although it's not telling us why...). This is the code in the `Service` construct that is responsible for creating the `httpsListener`: https://github.com/stratospheric-dev/cdk-constructs/blob/main/src/main/java/dev/stratospheric/cdk/Service.java#L72:

```java
Optional<String> httpsListenerArn = networkOutputParameters.getHttpsListenerArn();
    CfnListenerRule httpsListenerRule = null;
    if (httpsListenerArn.isPresent()) {
      httpsListenerRule = CfnListenerRule.Builder.create(this, "httpsListenerRule")
        .actions(singletonList(actionProperty))
        .conditions(singletonList(condition))
        .listenerArn(httpsListenerArn.get())
        .priority(1)
        .build();
    }
```

Since you left the parameter `sslCertificateArn` away when creating the `Network` construct, I would expect `httpsListenerArn.isPresent()` to return `false`. However, it seems that it's going into the if-block, after all.

Can you look up the value of the parameter `<environment>.Network.httpsListenerArn` in the AWS console? You should find the parameter in the parameter store of the "SSM" service.

**Comment by alrawas** [comment]
Oh,
Yes I can, this is the value ( I replaced accountId digits with dashes for privacy )
`arn:aws:elasticloadbalancing:us-east-1:------------:listener/app/STAGING-loadbalancer/7d2a243ae9596d21/b8fd9db1186a62f7`

**Comment by thombergs** [comment]
@alrawas is that really the `httpsListener` parameter or is it the `httpListener` parameter? 

According to this code in the `Network` construct, the `httpsListener` parameter shouldn't be stored when you don't provide the `sslCertificateArn`:

```java
if (this.httpsListener != null) {
    StringParameter httpsListener = StringParameter.Builder.create(this, "httpsListener")
      .parameterName(createParameterName(environmentName, PARAMETER_HTTPS_LISTENER))
      .stringValue(this.httpsListener.getListenerArn())
      .build();
  } else {
    StringParameter httpsListener = StringParameter.Builder.create(this, "httpsListener")
      .parameterName(createParameterName(environmentName, PARAMETER_HTTPS_LISTENER))
      .stringValue("null")
      .build();
  }
```

If it's really the `httpsListener` parameter and you really haven't provided the `sslCertificateArn`, you've found a bug :). Let me know if that really is the case, then I'll try to reproduce it.

**Comment by alrawas** [comment]
@thombergs Ouch, my bad!
The value I posted was the httpListener just like you said.
The value of <environment>-Network-httpsListenerArn is `null` indeed.

**Comment by alrawas** [comment]
More information:
in `stratospheric/chapters/chapter-6/cdk/src/main/java/dev/stratospheric/todoapp/cdk/ServiceApp.java`
just after this line:
```
Network.NetworkOutputParameters networkOutputParameters = Network
    .getOutputParametersFromParameterStore(serviceStack, applicationEnvironment.getEnvironmentName());
```
I logged `System.out.println(networkOutputParameters.getHttpsListenerArn().orElse("httpsListenerArn is null"));`
I got the following output: `${Token[TOKEN.23]}` and not null.

**Comment by thombergs** [comment]
Hey @alrawas,

I could reproduce this now, and I think I know the issue.

We fell for the synth time / deploy time issue. As you mentioned, the parameter is just a token at synth time, i.e. a placeholder for the real value that is loaded from the parameter store at deploy time. And since the placeholder is an actual value, this code in the `Service` construct doesn't work as expected:

```java
if (httpsListenerArn.isPresent()) {
  // ... create httpsListenerRule
}
```

It will try to create the `httpsListenerRule` even though the parameter in the parameter store is `null`. I'll try to fix this later today, stay tuned.

**Comment by alrawas** [comment]
closed by mistake

**Comment by thombergs** [comment]
@alrawas 

Alright, I found and fixed the issue in [this commit](https://github.com/stratospheric-dev/cdk-constructs/commit/901c2b75294acb4d551ddc839b9f1c28bb68026f).

I replaced the if statement with a CloudFormation condition that is evaluated at deploy time (instead of the if statement being evaluated at synthesis time).

I released a new version of the `cdk-constructs` library with the fix. If you update the dependency to `0.0.24`, it should work for you, too, hopefully :). It might take a few hours until the new version is available on Maven Central, though. 

Let me know if it works. The code path without the `sslCertificateArn` wasn't properly tested, obviously ... sorry that you had to be our guinea pig here.
